### PR TITLE
[libSyntax] Incremental Syntax Parsing

### PR DESCRIPTION
**Pull request by ahoppen**
This is a work in progress PR to add incremental syntax parsing to the Swift compiler. It accompanies the proposal [here](https://gist.github.com/ahoppen/3ae1a6cd64e558710a4afcd372e8fdc4) and discussion thread [here](https://forums.swift.org/t/incremental-syntax-parsing/12368).

### CODE REVIEW

File: lib/Parse/SyntaxParsingCache.cpp [hunk 1]
```diff
@@ -15,6 +15,45 @@
 using namespace swift;
 using namespace swift::syntax;
 
+ArrayRef<TriviaPiece> getNextChildsLeadingTrivia(const Syntax *Node) {
```
Comment (harlanhaskins) [inline lib/Parse/SyntaxParsingCache.cpp]: Would be nice to make this `getNextChild`, because that's a traversal I've found very necessary.

Comment (ahoppen) [reply lib/Parse/SyntaxParsingCache.cpp]: Sounds good. I will keep it in mind when I refactor these methods to provide a (hopefully faster) cache based lookup.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: include/swift/Syntax/RawSyntax.h [hunk 2]
```diff
@@ -434,6 +442,28 @@ class RawSyntax final
     return getLayout()[Index];
   }
 
+  /// Return the number of bytes this node takes when spelled out in the source
+  size_t getTextLength() {
```
Comment (nkcsgexi) [inline include/swift/Syntax/RawSyntax.h]: can we re-use `accumulateAbsolutePosition` to calculate the length?

File: include/swift/Parse/SyntaxParsingCache.h [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  void addEdit(size_t Start, size_t End, size_t ReplacementLength) {
+    Edits.push_back({Start, End, ReplacementLength});
+  }
+
+  /// Check if a syntax node of the given kind at the given position can be
+  /// reused for a new syntax tree.
+  llvm::Optional<Syntax> lookUp(size_t NewPosition, SyntaxKind Kind);
+
+  /// Turn recording of reused ranges on
+  void recordReuseInformation() { RecordReuseInformation = true; }
```
Comment (nkcsgexi) [inline include/swift/Parse/SyntaxParsingCache.h]: A better name is `setRecordReuseInformation()`.

File: include/swift/Parse/SyntaxParsingContext.h [hunk 4]
```diff
@@ -112,6 +113,9 @@ class alignas(1 << SyntaxAlignInBits) SyntaxParsingContext {
   // Reference to the
   SyntaxParsingContext *&CtxtHolder;
 
+  /// A cache of nodes that can be reused when creating the current syntax tree
+  SyntaxParsingCache *SyntaxParsingCache = nullptr;
```
Comment (nkcsgexi) [inline include/swift/Parse/SyntaxParsingContext.h]: We could move this instance to `RootContextData`. It will make it clear that all `SyntaxParsingContext` share a single cache entry. Same with `Arena`.

File: lib/Parse/SyntaxParsingContext.cpp [hunk 5]
```diff
@@ -37,13 +37,34 @@ SyntaxParsingContext::SyntaxParsingContext(SyntaxParsingContext *&CtxtHolder,
...
+  if (!SyntaxParsingCache) {
+    // We don't have a cache, so there's nothing to look up
+    return 0;
+  }
+  auto CacheLookup = SyntaxParsingCache->lookUp(LexerOffset, SynKind);
+  if (!CacheLookup) {
+    return 0;
+  }
+  // Disable this node and push the cached nodes into the storage
+  disable();
```
Comment (nkcsgexi) [inline lib/Parse/SyntaxParsingContext.cpp]: Instead of calling `disable()`, can we define another `AccumulationMode` called `LoadedFromCache` to be more specific? Probably, we can assert `LoadedFromCache` context doesn't spawn any child context.

**Review by nkcsgexi — COMMENTED** [review]
Looking good overall! Some comments so far.

File: include/swift/Syntax/SyntaxKind.h.gyb [hunk 6]
```diff
@@ -71,6 +71,8 @@ bool isTokenKind(SyntaxKind Kind);
 
 bool isUnknownKind(SyntaxKind Kind);
 
+bool shallBeOmittedWhenNoChildren(SyntaxKind Kind);
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply include/swift/Syntax/SyntaxKind.h.gyb]: I didn't want to create a gyb file in the parser for this. It is currently still nice and gyb-free.

Maybe we could add this function to the `swift` namespace instead of the `swift::syntax` one and call it `parserShallOmitIfNoChildern`

[the remark this answers is not in the document]
Comment (nkcsgexi) [reply include/swift/Syntax/SyntaxKind.h.gyb]: yeah, it works too, as long as we make it explicit that it's a parser thing rather than for syntax nodes in general.

File: include/swift/Syntax/RawSyntax.h (hunk 2 above)
Comment (ahoppen) [reply include/swift/Syntax/RawSyntax.h]: I'm using `accumulateAbsolutePosition` now to compute the length of a token. For all other nodes I'm still computing the value while constructing the tree because it is faster.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: lib/Parse/Parser.cpp [hunk 7]
```diff
@@ -709,6 +709,23 @@ void Parser::skipUntilConditionalBlockClose() {
...
+    return false;
+  }
+  unsigned LexerOffset =
+      SourceMgr.getLocOffsetInBuffer(Tok.getLoc(), L->getBufferID());
+  if (auto TextLength = SyntaxContext->loadFromCache(
+          LexerOffset - LeadingTrivia.getTextLength())) {
+    L->advanceTo(Tok, LeadingTrivia, TextLength);
+    L->lex(Tok, LeadingTrivia, TrailingTrivia);
+    return true;
+  }
```
Comment (rintaro) [inline lib/Parse/Parser.cpp]: I feel this needlessly complicated.
How about:
```cpp
  unsigned Offset = SourceMgr.getLocOffsetInBuffer(Tok.getLoc(), L->getBufferID());
  Offset -= LeadingTrivia.getTextLength();
  if (auto TextLength = SyntaxContext->loadFromCache(Offset)) {
    L->resetToOffset(Offset + TextLength);
    L->lex(Tok, LeadingTrivia, TrailingTrivia);
    return true;
  }
```
where `Lexer:: resetToOffset` is just:
```swift
Lexer::resetToOffset(size_t Offset) {
  assert(BufferStart + Offset <= BufferEnd);
  CurPtr = BufferStart + Offset;

  lexImpl();
}
```

File: lib/Parse/ParseDecl.cpp [hunk 8]
```diff
@@ -3349,6 +3352,8 @@ ParserResult<PoundDiagnosticDecl> Parser::parseDeclPoundDiagnostic() {
 }
 
 ParserStatus Parser::parseLineDirective(bool isLine) {
+  SyntaxParsingContext PoundSourceLocation(SyntaxContext,
+                                           SyntaxKind::PoundSourceLocation);
```
Comment (rintaro) [inline lib/Parse/ParseDecl.cpp]: Good catch!
Could you separate `#sourceLocation` parsing out of this PR if possible?

File: test/lit.cfg [hunk 9]
```diff
@@ -721,7 +722,7 @@ if run_vendor == 'apple':
     config.target_sdk_name = xcrun_sdk_name
     config.target_ld = "%s ld -L%r" % (xcrun_prefix, make_path(test_resource_dir, config.target_sdk_name))
     config.target_swift_frontend = (
-        "%s -frontend %s -sdk %s %s %s" %
+        "%s -frontend %s -sdk '%s' %s %s" %
```
Comment (rintaro) [inline test/lit.cfg]: Unrelated change?

**Review by rintaro — COMMENTED** [review]
Looks pretty good overall.
Some comments, nothing critical.

File: lib/Parse/SyntaxParsingCache.cpp [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+  // StructDecl inside and `private struc Foo {}` parses as two CodeBlockItems
+  // one for `private` and one for `struc Foo {}`
+  size_t NextLeafNodeLength = 0;
+  if (auto NextNode = Node.getData().getNextNode()) {
+    auto NextLeafNode = NextNode->getFirstLeafNode();
+    assert(NextLeafNode->getRaw()->isToken() && "Leaf node that is no token?");
+    auto NextRawNode = NextLeafNode->getRaw();
+    NextLeafNodeLength += NextRawNode->getTextLength();
+    for (auto TriviaPiece : NextRawNode->getLeadingTrivia()) {
+      NextLeafNodeLength += TriviaPiece.getTextLength();
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply lib/Parse/SyntaxParsingCache.cpp]: This is a bug. I fixed it in the syntax colouring branch but missed to cherry-pick it over. Done that now.

File: include/swift/Parse/SyntaxParsingContext.h [hunk 11]
```diff
@@ -174,20 +183,41 @@ class alignas(1 << SyntaxAlignInBits) SyntaxParsingContext {
 
   ~SyntaxParsingContext();
 
+  /// Try loading the current node from the \c SyntaxParsingCache by looking up
+  /// if an unmodified node exists at \c LexerOffset of the same kind. If a node
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply include/swift/Parse/SyntaxParsingContext.h]: Thanks. I didn't know that.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by harlanhaskins** [comment]
🎉🎉🎉
